components/gitpod-db/src/typeorm/synchronizer.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable } from "inversify";
+import { TypeORM } from "./typeorm";
+
+@injectable()
+export class Synchronizer {
+    @inject(TypeORM) typeORM: TypeORM;
+
+    /**
+     * runs the given function with a distributed lock based on mysql's GET_LOCK
+     */
+    async synchronized<T>(lockKey: string, component: string, fn: () => Promise<T>): Promise<T> {
+        // we use a query runner because that will use the same connection exclusively.
+        // GET_LOCK is per connection, so we need to make sure we're using the same connection
+        // note that calling this method will block the connection for the duration of the function, so if we exceed the pool with lots of concurrent calls, we'll run into trouble
```

Comment:
Since the synchronizer is in singleton scope we could have an additional wait block in this component for concurrent calls that is lower than the connection pool size.